File: Auto_Use/web/controller/scratchpad/service.rs
// Copyright 2026 Ashish Yadav — Auto-Use

//! The scratchpad — the agent's own numbered notes, persisted per run.
//!
//! On-disk storage stays "milestone/milestone.md" to preserve existing data
//! and avoid a scratchpad/scratchpad/scratchpad.md collision with the parent
//! directory.

use std::path::{Path, PathBuf};

pub struct ScratchpadService {
    pub scratchpad_dir: PathBuf,
    pub scratchpad_file: PathBuf,
}

impl ScratchpadService {
    /// `scratchpad_base` is Auto_Use/web/scratchpad, or a per-session
    /// subdirectory of it when agents run in parallel.
    pub fn new(scratchpad_base: &Path) -> std::io::Result<Self> {
        let scratchpad_dir = scratchpad_base.join("milestone");
        std::fs::create_dir_all(&scratchpad_dir)?;
        let scratchpad_file = scratchpad_dir.join("milestone.md");
        Ok(ScratchpadService { scratchpad_dir, scratchpad_file })
    }

    /// Append a scratchpad entry with sequential numbering. Returns whether
    /// the write landed — a failure becomes a failed tool result, never a
    /// panic.
    pub fn append_scratchpad(&self, content: &str) -> bool {
        let existing_count = std::fs::read_to_string(&self.scratchpad_file)
            .map(|text| text.lines().filter(|ln| !ln.trim().is_empty()).count())
            .unwrap_or(0);
        let entry = format!("{}. {content}\n", existing_count + 1);
        use std::io::Write;
        std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.scratchpad_file)
            .and_then(|mut f| f.write_all(entry.as_bytes()))
            .is_ok()
    }

    /// Read current scratchpad content.
    pub fn read_scratchpad(&self) -> String {
        std::fs::read_to_string(&self.scratchpad_file)
            .map(|s| s.trim().to_string())
            .unwrap_or_default()
    }

    /// Clear the scratchpad file.
    pub fn clear_scratchpad(&self) -> bool {
        if self.scratchpad_file.exists() {
            std::fs::remove_file(&self.scratchpad_file).is_ok()
        } else {
            true
        }
    }
}

File: Auto_Use/ios/controller/scratchpad/service.py
# Copyright 2026 Cursortouch — Auto-Use

import os
import logging

# Configure logger
logger = logging.getLogger(__name__)

class ScratchpadService:
    def __init__(self):
        """Initialize the Scratchpad Service"""
        current_dir = os.path.dirname(os.path.abspath(__file__))
        # Go up two levels from Auto_Use/ios/controller/scratchpad/ to reach ios/
        ios_dir = os.path.dirname(os.path.dirname(current_dir))

        # On-disk storage kept as "milestone/milestone.md" to avoid the
        # scratchpad/scratchpad/scratchpad.md collision with the parent dir.
        # Parallel simulator tasks each get their own scratchpad/<session>/
        # subtree (AUTOUSE_IOS_SESSION), so they never overwrite each other.
        session = os.environ.get("AUTOUSE_IOS_SESSION") or ""
        self.scratchpad_dir = os.path.join(ios_dir, "scratchpad", session, "milestone") \
            if session else os.path.join(ios_dir, "scratchpad", "milestone")
        self.scratchpad_file = os.path.join(self.scratchpad_dir, "milestone.md")

        # Create scratchpad directory if it doesn't exist
        self._ensure_scratchpad_directory()

    def _ensure_scratchpad_directory(self):
        """Create scratchpad directory if it doesn't exist"""
        try:
            os.makedirs(self.scratchpad_dir, exist_ok=True)
        except Exception as e:
            logger.error(f"Error creating scratchpad directory: {str(e)}")
            raise

    def append_scratchpad(self, scratchpad_content):
        """Append a scratchpad entry with sequential numbering"""
        try:
            existing_count = 0
            if os.path.exists(self.scratchpad_file):
                with open(self.scratchpad_file, "r", encoding="utf-8") as f:
                    for line in f:
                        if line.strip():
                            existing_count += 1

            next_number = existing_count + 1

            with open(self.scratchpad_file, "a", encoding="utf-8") as f:
                f.write(f"{next_number}. {scratchpad_content}\n")

            return True

        except Exception as e:
            logger.error(f"Error appending scratchpad entry: {str(e)}")
            return False

    def read_scratchpad(self):
        """Read current scratchpad content"""
        try:
            if os.path.exists(self.scratchpad_file):
                with open(self.scratchpad_file, "r", encoding="utf-8") as f:
                    return f.read().strip()
            return ""
        except Exception as e:
            logger.error(f"Error reading scratchpad: {str(e)}")
            return ""

    def clear_scratchpad(self):
        """Clear the scratchpad file"""
        try:
            if os.path.exists(self.scratchpad_file):
                os.remove(self.scratchpad_file)
            return True
        except Exception as e:
            logger.error(f"Error clearing scratchpad: {str(e)}")
            return False
